src/Features/Core/Portable/ChangeSignature/AbstractChangeSignatureService.cs

```diff
@@ -665,151 +676,227 @@ protected static int GetParameterIndex<TNode>(SeparatedSyntaxList<TNode> paramet
+                    }
+
                     newParameters.Add(newParameter);
+                    numAddedParameters++;
                 }
             }
 
-            int numSeparatorsToSkip;
-            if (originalParameters.Count == 0)
+            // (a,b,c)
+            // Adding X parameters, need to add X separators.
+            var numSeparatorsToSkip = originalParameters.Length - reorderedParameters.Length;
+
+            if (originalParameters.Length == 0)
             {
                 // () 
                 // Adding X parameters, need to add X-1 separators.
-                numSeparatorsToSkip = originalParameters.Count - reorderedParameters.Count + 1;
-            }
-            else
-            {
-                // (a,b,c)
-                // Adding X parameters, need to add X separators.
-                numSeparatorsToSkip = originalParameters.Count - reorderedParameters.Count;
+                numSeparatorsToSkip++;
             }
 
-            return (newParameters, GetSeparators(list, numSeparatorsToSkip));
+            return (newParameters.ToImmutable(), GetSeparators(list, numSeparatorsToSkip));
         }
 
-        protected List<SyntaxToken> GetSeparators<T>(SeparatedSyntaxList<T> arguments, int numSeparatorsToSkip = 0) where T : SyntaxNode
+        protected ImmutableArray<SyntaxToken> GetSeparators<T>(SeparatedSyntaxList<T> arguments, int numSeparatorsToSkip) where T : SyntaxNode
         {
-            var separators = new List<SyntaxToken>();
+            var separators = ImmutableArray.CreateBuilder<SyntaxToken>();
 
             for (int i = 0; i < arguments.SeparatorCount - numSeparatorsToSkip; i++)
             {
-                if (i >= arguments.SeparatorCount)
-                {
-                    separators.Add(Generator.CommaTokenWithElasticSpace());
-                }
-                else
-                {
-                    separators.Add(arguments.GetSeparator(i));
-                }
+                separators.Add(i < arguments.SeparatorCount
+                    ? arguments.GetSeparator(i)
+                    : CommaTokenWithElasticSpace());
             }
 
-            return separators;
+            return separators.ToImmutable();
         }
 
         protected abstract SyntaxGenerator Generator { get; }
+        protected abstract ISyntaxFacts SyntaxFacts { get; }
 
-        protected SeparatedSyntaxList<SyntaxNode> AddNewArgumentsToList(
-          SeparatedSyntaxList<SyntaxNode> newArguments,
-          SignatureChange signaturePermutation,
-          bool isReducedExtensionMethod)
+        protected virtual SeparatedSyntaxList<SyntaxNode> AddNewArgumentsToList(
+            ISymbol declarationSymbol,
+            SeparatedSyntaxList<SyntaxNode> newArguments,
+            SignatureChange signaturePermutation,
+            bool isReducedExtensionMethod,
+            bool isParamsArrayExpanded,
+            bool generateAttributeArguments = false)
```

Comment:
nit: make this non-optional.  i want callers to ahve to think about what to pass here.

Reply:
Done.